libkpod/ps.go

```diff
@@ -0,0 +1,192 @@
+func findContainer(containers []*oci.Container, ref string) (*oci.Container, error) {
+	for _, ctr := range containers {
+		if strings.HasPrefix(ctr.ID(), ref) || ctr.Name() == ref {
+			return ctr, nil
+		}
+	}
+	return nil, errors.Errorf("could not find container")
+}
+
+// matchesFilter checks if a container matches any of the filter parameters
+func matchesFilter(server *ContainerServer, container *oci.Container, params *FilterParamsPS) bool {
+	if params == nil {
+		return true
+	}
+
+	ctrData, err := server.GetContainerData(container.ID(), true)
+	if err != nil {
+		return false
+	}
+	if params.id != "" && !matchesID(ctrData, params.id) {
+		return false
+	}
+	if params.name != "" && !matchesName(ctrData, params.name) {
+		return false
+	}
+	if !params.before.IsZero() && !matchesBeforeContainer(ctrData, params.before) {
+		return false
+	}
+	if !params.since.IsZero() && !matchesSinceContainer(ctrData, params.since) {
+		return false
+	}
+	if params.exited > -1 && !matchesExited(ctrData, params.exited) {
+		return false
+	}
+	if params.status != "" && !matchesStatus(ctrData, params.status) {
+		return false
+	}
+	if params.ancestor != "" && !matchesAncestor(ctrData, params.ancestor) {
+		return false
+	}
+	if params.label != "" && !matchesLabel(ctrData, params.label) {
+		return false
+	}
+	if params.volume != "" && !matchesVolume(ctrData, params.volume) {
+		return false
+	}
+	return true
+}
+
+// GetContainersMatchingFilter returns a slice of all the containers that match the provided filter parameters
+func GetContainersMatchingFilter(server *ContainerServer, filter *FilterParamsPS) ([]*ContainerData, error) {
+	containers := server.ListContainers()
+	var filteredCtrs []*ContainerData
+	for _, ctr := range containers {
+		if filter == nil || matchesFilter(server, ctr, filter) {
+			ctrData, err := server.GetContainerData(ctr.ID(), true)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error getting container data")
+			}
+			filteredCtrs = append(filteredCtrs, ctrData)
+		}
+	}
+	return filteredCtrs, nil
+}
+
+// matchesID returns true if the id's match
+func matchesID(ctrData *ContainerData, id string) bool {
+	return strings.HasPrefix(ctrData.ID, id)
+}
+
+// matchesBeforeContainer returns true if the container was created before the filter image
+func matchesBeforeContainer(ctrData *ContainerData, beforeTime time.Time) bool {
+	return ctrData.State.Created.Before(beforeTime)
+}
+
+// matchesSincecontainer returns true if the container was created since the filter image
+func matchesSinceContainer(ctrData *ContainerData, sinceTime time.Time) bool {
+	return ctrData.State.Created.After(sinceTime)
+}
+
+// matchesLabel returns true if the container label matches that of the filter label
+func matchesLabel(ctrData *ContainerData, label string) bool {
+	pair := strings.SplitN(label, "=", 2)
+	if val, ok := ctrData.Labels[pair[0]]; ok {
+		if len(pair) == 2 && val == pair[1] {
+			return true
```

Comment:
It is possible for the user to filter using a key only or key and value: `--filter label=key` or `--filter label=key=value`. So returning true also if only a key is given to the filter and there is a match.